streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java

```diff
@@ -103,8 +104,12 @@ public void init(final ProcessorContext context,
 
     @SuppressWarnings("unchecked")
     void initStoreSerde(final ProcessorContext context) {
+        final String storeName = name();
+        final String changelogTopic = ProcessorContextUtils.changelogFor(context, storeName);
         serdes = new StateSerdes<>(
-            ProcessorStateManager.storeChangelogTopic(context.applicationId(), name()),
+             changelogTopic != null ?
+                changelogTopic :
+                ProcessorStateManager.storeChangelogTopic(context.applicationId(), storeName),
```

Comment:
I think we still need the if-then-else. `ProcessorContextUtils.changelogFor()` checks that the processor context is of type `InternalProcessorContext` to be able to call `changelogFor()` on it. It does not consider the case where no changelog exists (i.e., `changelogFor()` returns `null`). The `null` check is done here. Fallback is in both cases `ProcessorStateManager.storeChangelogTopic()`.

Reply:
Ah thanks. I missed this case.

However, should we move both `null` checks into `ProcessorContextUtils.changelogFor()` for this case? It seem, we do the same "outer" `null`-check each time we call the method, so why not do it at a single place in the code?

Reply:
I would not put any code that is not related to casts of `ProcessorContext` into `ProessorContextUtils`. I think the goal of `ProessorContextUtils` is to contain all code of which we want to get rid of in the future once the casts are fixed.

We could move the `null` check into the constructor of `StateSerde` since we do also there a `null` check.